lua/wire/stools/expression2.lua

```diff
@@ -259,156 +273,127 @@ if SERVER then
 			return
 		end
 
-		if not IsValid(ply) or not ply:IsPlayer() then -- wtf
+		if not IsValid(ply) or not ply:IsPlayer() then
 			error("Invalid player entity (wtf??). This should never happen. " .. tostring(ply), 0)
 		end
 
 		local main, includes = targetEnt:GetCode()
-		if not includes or not next(includes) then -- There are no includes
-			local datastr = WireLib.von.serialize({ { targetEnt.name, main } })
-			local numpackets = math.ceil(#datastr / 64000)
-
-			local n = 0
-			for i = 1, #datastr, 64000 do
-				timer.Simple( n, function()
-					if not IsValid( targetEnt ) then return end
-					net.Start("wire_expression2_download")
-						net.WriteEntity(targetEnt)
-						net.WriteBit(uploadandexit or false)
-						net.WriteUInt(numpackets, 16)
-						local data = datastr:sub(i, i + 63999)
-						net.WriteUInt(#data, 32)
-						net.WriteData(data, #data)
-					net.Send(ply)
-				end)
-				n = n + 1
-			end
-		elseif not wantedfiles then
-			local data = {}
-			for k, v in pairs(includes) do
-				data[#data + 1] = k
+		if (not includes or not next(includes)) and not wantedfiles then
+			local data = { { targetEnt.name, main } }
+			local datastr = WireLib.von.serialize(data)
+
+			net.Start("wire_expression2_download")
+				net.WriteEntity(targetEnt)
+				net.WriteBit(uploadandexit or false)
+				net.WriteStream(datastr)
+			net.Send(ply)
+			return
+		end
+
+		if not wantedfiles then
+			local fileList = {}
+			for filename, _ in pairs(includes) do
+				fileList[#fileList + 1] = filename
 			end
 
-			local datastr = WireLib.von.serialize(data)
+			local datastr = WireLib.von.serialize(fileList)
 			net.Start("wire_expression2_download_wantedfiles_list")
-			net.WriteEntity(targetEnt)
-			net.WriteBit(uploadandexit or false)
-			net.WriteUInt(#datastr, 32)
-			net.WriteData(datastr, #datastr)
+				net.WriteEntity(targetEnt)
+				net.WriteBit(uploadandexit or false)
+				net.WriteStream(datastr)
 			net.Send(ply)
+
 			targetEnt.DownloadAllowedPlayers = targetEnt.DownloadAllowedPlayers or {}
 			targetEnt.DownloadAllowedPlayers[ply] = true
-			timer.Simple(60, function() -- make permissions timeout after 60 seconds
-				if not IsValid(targetEnt) then return end
-				if not targetEnt.DownloadAllowedPlayers then return end
-				if not targetEnt.DownloadAllowedPlayers[ply] then return end
-				targetEnt.DownloadAllowedPlayers[ply] = nil
-				if table.IsEmpty(targetEnt.DownloadAllowedPlayers) then
-					targetEnt.DownloadAllowedPlayers = nil
+			timer.Simple(60, function()
+				if IsValid(targetEnt) and targetEnt.DownloadAllowedPlayers then
+					targetEnt.DownloadAllowedPlayers[ply] = nil
+					if table.IsEmpty(targetEnt.DownloadAllowedPlayers) then
+						targetEnt.DownloadAllowedPlayers = nil
+					end
 				end
 			end)
```

Comment:
I'll be looking at this PR again for a bit more now that I got some time.

This timer.Simple needs to be removed and `targetEnt.DownloadAllowedPlayers[ply] = true` needs to be `targetEnt.DownloadAllowedPlayers[ply] = CurTime()+60` and the check for DownloadAllowedPlayers comparing against CurTime.